core/drivers/versal_net_nvm.c

```diff
@@ -0,0 +1,1060 @@
+
+	if (!buf)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ret = versal_efuse_read_cache(EFUSE_CACHE_PUF_ECC_CTRL_OFFSET, 1,
+				      &puf_ctrl, sizeof(puf_ctrl));
+	if (ret)
+		return ret;
+
+	/*
+	 * Some fuses have moved from PUF_ECC_CTRL to SECURITY_CTRL
+	 */
+	ret = versal_efuse_read_cache(EFUSE_CACHE_SEC_CTRL_OFFSET, 1,
+				      &sec_ctrl, sizeof(sec_ctrl));
+	if (ret)
+		return ret;
+
+	buf->puf_regen_dis = (puf_ctrl & BIT(31)) >> 31;
+	buf->puf_hd_invalid = (puf_ctrl & BIT(30)) >> 30;
+	buf->puf_test2_dis = (puf_ctrl & BIT(29)) >> 29;
+	buf->puf_dis = (sec_ctrl & BIT(18)) >> 18;
+	buf->puf_syn_lk = (sec_ctrl & BIT(16)) >> 16;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result versal_efuse_read_puf(struct versal_efuse_puf_header *buf)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (!buf)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ret = versal_efuse_read_puf_sec_ctrl(&buf->sec_ctrl);
+	if (ret)
+		return ret;
+
+	ret = versal_efuse_read_cache(EFUSE_CACHE_SEC_CTRL_OFFSET, 1,
+				      &buf->aux, sizeof(uint32_t));
+	if (ret)
+		return ret;
+
+	buf->aux &= GENMASK_32(23, 0);
+
+	ret = versal_efuse_read_cache(EFUSE_CACHE_PUF_CHASH_OFFSET, 1,
+				      &buf->chash, sizeof(uint32_t));
+	if (ret)
+		return ret;
+
+	return versal_efuse_read_cache(EFUSE_CACHE_PUF_SYN0_OFFSET,
+				      PUF_SYN_DATA_WORDS, buf->efuse_syn_data,
+				      PUF_SYN_DATA_WORDS * NVM_WORD_LEN);
+}
+
+TEE_Result
+versal_efuse_read_puf_as_user_fuse(struct versal_efuse_puf_user_fuse
+				   *p __unused)
+{
+	return TEE_ERROR_NOT_IMPLEMENTED;
+}
+
+TEE_Result versal_efuse_write_user_data(uint32_t *buf __unused,
+					size_t len __unused,
+					uint32_t first __unused,
+					size_t num __unused)
+{
+	return TEE_ERROR_NOT_IMPLEMENTED;
+}
+
+#define EFUSE_ENV_DIS_FLAG		0
+
+#define EFUSE_AES_KEY_ID		0
+#define EFUSE_USER_KEY0_ID		1
+#define EFUSE_USER_KEY1_ID		2
+
+static TEE_Result do_write_efuses_buffer(enum versal_nvm_api_id id,
+					 uint16_t type, uint32_t *buf,
+					 size_t len)
+{
+	struct versal_ipi_cmd cmd = { };
+	struct versal_mbox_mem p = { };
+	TEE_Result ret = TEE_SUCCESS;
+	uint32_t a = 0;
+	uint32_t b = 0;
+
+	ret = versal_mbox_alloc(len, buf, &p);
+	if (ret)
+		return ret;
+
+	reg_pair_from_64(virt_to_phys(p.buf), &b, &a);
+
+	cmd.data[0] = NVM_API_ID(id);
+	cmd.data[1] = (type << 16) | EFUSE_ENV_DIS_FLAG;
+	cmd.data[2] = a;
+	cmd.data[3] = b;
+
+	cmd.ibuf[0].mem = p;
+
+	ret = versal_mbox_notify_pmc(&cmd, NULL, NULL);
+
+	versal_mbox_free(&p);
+	return ret;
+}
+
+static TEE_Result do_write_efuses_value(enum versal_nvm_api_id id, uint32_t val)
+{
+	struct versal_ipi_cmd cmd = { };
+
+	cmd.data[0] = NVM_API_ID(id);
+	cmd.data[1] = EFUSE_ENV_DIS_FLAG;
+	cmd.data[2] = val;
+
+	return versal_mbox_notify_pmc(&cmd, NULL, NULL);
+}
+
+static TEE_Result do_write_efuses(enum versal_nvm_api_id id)
+{
+	return do_write_efuses_value(id, 0);
+}
+
+TEE_Result versal_efuse_write_aes_keys(struct versal_efuse_aes_keys *keys)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	TEE_Result res = TEE_SUCCESS;
```

Comment:
`res` and `ret` variable names are quite alike. Prefer `res` and `res2` for example.

Reply:
Fixed in the latest push.